SPECS/SymCrypt/SymCrypt.spec

```diff
@@ -40,26 +42,47 @@ A core cryptographic library written by Microsoft
 %endif
 
 %prep
-%setup -q
-%setup -q -a 1
+%autosetup -a 1 -p1
 # Create a symbolic link as if jitterentropy-library has been pulled in as git submodule
 rm -rf 3rdparty/jitterentropy-library
 ln -s ../jitterentropy-library-3.3.1 3rdparty/jitterentropy-library
+# Copy the custom find-debuginfo script
+cp %{SOURCE2} .
 
 %build
 SYMCRYPT_BRANCH=main \
 SYMCRYPT_COMMIT_HASH=a84ffe1 \
 SYMCRYPT_COMMIT_TIMESTAMP=2024-01-26T22:00:47-08:00 \
 cmake   -S . -B bin \
         -DSYMCRYPT_TARGET_ARCH=%{symcrypt_arch} \
-        -DCMAKE_BUILD_TYPE=Release \
+        -DSYMCRYPT_STRIP_BINARY=OFF \
+        -DSYMCRYPT_FIPS_POSTPROCESS=OFF \
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo \
         -DCMAKE_C_COMPILER=%{symcrypt_cc} \
         -DCMAKE_CXX_COMPILER=%{symcrypt_cxx} \
         -DCMAKE_C_FLAGS="%{symcrypt_c_flags}" \
         -DCMAKE_CXX_FLAGS="-Wno-unused-but-set-variable"
 
 cmake --build bin
 
+# Override the default find-debuginfo script to our own custom one, which is modified
+# to allow us to keep symbols.
+# Also add custom options to the call to find-debuginfo.
+%define __find_debuginfo ./find-debuginfo
+%define _find_debuginfo_opts --keep-symbol SymCryptVolatileFipsHmacKey --keep-symbol SymCryptVolatileFipsHmacKeyRva --keep-symbol SymCryptVolatileFipsBoundaryOffset --keep-symbol SymCryptVolatileFipsHmacDigest
+
+# Override the default to allow us to do custom fips post-processing after debug info/stripping is done.
+# The post-processing script writes the modified file to the same location as the original file, which
+# is subject to default permissions, so we need to set permissions manually after the script.
+%define __spec_install_post \
+    %{?__debug_package:%{__debug_install_post}} \
+    %{__arch_install_post} \
+    %{__os_install_post} \
+    mkdir -p "bin/module/generic/processing" \
+    python3 "scripts/process_fips_module.py" "%{buildroot}%{_libdir}/libsymcrypt.so.%{version}" --processing-dir "bin/module/generic/processing" --debug \
+    chmod 755 "%{buildroot}%{_libdir}/libsymcrypt.so.%{version}" \
+%{nil}
```

Comment:
Small request to try something out. Since we're only appending to the original `__spec_install_post`, I'm curious if we could do this to make it clearer what is the new part and that we're not skipping any original steps:
```
%define __spec_install_post \
    %{?__spec_install_post} \
    mkdir -p "bin/module/generic/processing" \
    python3 "scripts/process_fips_module.py" "%{buildroot}%{_libdir}/libsymcrypt.so.%{version}" --processing-dir "bin/module/generic/processing" --debug \
    chmod 755 "%{buildroot}%{_libdir}/libsymcrypt.so.%{version}" \
tobiasb-ms marked this conversation as resolved.
%{nil}
```

That way, if `__spec_install_post` gets updated, we also won't miss out on the update.

Reply:
That's a great idea, assuming it works. I'll experiment.

Reply:
I tried this and it gets an infinite recursion issue. Doesn't seem like it will work. Too bad.